--- a/netflix/src/platform/gibbon/CMakeLists.txt	2015-01-30 16:38:41.625905063 +0100
+++ b/netflix/src/platform/gibbon/CMakeLists.txt	2015-01-30 16:02:55.725943492 +0100
@@ -257,6 +257,7 @@
 # FreeType
 # -----------------------------------------------------------------------------------------------------------
 if (NOT HAVE_LIBFREETYPE AND (GIBBON_TEXT STREQUAL "auto" OR GIBBON_TEXT STREQUAL "freetype"))
+    set(CMAKE_REQUIRED_INCLUDES "$ENV{STAGING_DIR}/usr/include/freetype2")
     set(CMAKE_REQUIRED_LIBRARIES "freetype;z;png")
     set(LIBFREETYPE_TEST "
         #include <ft2build.h>
@@ -587,6 +588,9 @@
 
 configure_file_if_different("config.h.in" "${CMAKE_CURRENT_BINARY_DIR}/include/gibbon/config.h")
 include_directories("${CMAKE_CURRENT_BINARY_DIR}/include")
+include_directories("$ENV{STAGING_DIR}/usr/include/directfb")
+include_directories("$ENV{STAGING_DIR}/usr/include/freetype2")
+link_directories("$ENV{STAGING_DIR}/usr/lib")
 
 message("=============================================================")
 message(" ")
@@ -798,8 +798,12 @@
     endif()
 
     get_environment_variable(DUMP_SYMS "Path to breakpad's DUMP_SYMS executable" "${TOOLCHAIN_DIRECTORY}/netflix/bin/dump_syms")
-    get_environment_variable(STRIP "Path to the strip executable" "strip")
-    get_environment_variable(OBJCOPY "Path to the objcopy executable" "objcopy")
+    message ("########################################################")
+    message ("LOOKING FOR -------[$ENV{BUILDROOT_TOOL_PREFIX}strip]------------")
+    message ("LOOKING FOR ------[$ENV{BUILDROOT_TOOL_PREFIX}objcopy]-----------")
+    message ("########################################################")
+    get_environment_variable(STRIP "Path to the strip executable" "$ENV{BUILDROOT_TOOL_PREFIX}strip")
+    get_environment_variable(OBJCOPY "Path to the objcopy executable" "$ENV{BUILDROOT_TOOL_PREFIX}objcopy")
 
     if(GIBBON_MODE STREQUAL "executable")
         set(APPLICATION_EXECUTABLE "netflix")

