--- a/netflix/gibbondir/dpi/reference/CMakeLists.txt	2015-10-08 12:31:50.847152241 +0200
+++ b/netflix/gibbondir/dpi/reference/CMakeLists.txt	2015-10-08 12:30:02.724924358 +0200
@@ -213,20 +213,17 @@
 else()
     add_definitions("-DBUILD_FFMPEG_AUDIO_DECODER")
     list(APPEND SOURCES x86/AudioDecoderNative.cpp)
-    set(CMAKE_C_FLAGS "-I${TOOLCHAIN_DIRECTORY}/netflix/include/ffmpeg2 ${CMAKE_C_FLAGS}")
-    set(CMAKE_CXX_FLAGS "-I${TOOLCHAIN_DIRECTORY}/netflix/include/ffmpeg2 ${CMAKE_CXX_FLAGS}")
-    #include_directories(BEFORE ${TOOLCHAIN_DIRECTORY}/netflix/include/ffmpeg2)
     list(APPEND LIBRARIES
-        ${TOOLCHAIN_DIRECTORY}/netflix/lib/ffmpeg2/libavcodec.a
-        ${TOOLCHAIN_DIRECTORY}/netflix/lib/ffmpeg2/libswresample.a
-        ${TOOLCHAIN_DIRECTORY}/netflix/lib/ffmpeg2/libavutil.a
+        libavcodec.a
+        libswresample.a
+        libavutil.a
     )
     if(EXISTS ${TOOLCHAIN_DIRECTORY}/netflix/lib/ffmpeg2/libx264.a) # rpi :(
         list(APPEND LIBRARIES ${TOOLCHAIN_DIRECTORY}/netflix/lib/ffmpeg2/libx264.a)
     endif()
     if(EXISTS ${TOOLCHAIN_DIRECTORY}/netflix/lib/dcv_sub.a) # Dolby library
         add_library(ddplus STATIC IMPORTED)
-        set_property(TARGET ddplus PROPERTY IMPORTED_LOCATION ${TOOLCHAIN_DIRECTORY}/netflix/lib/dcv_sub.a)
+        set_property(TARGET ddplus PROPERTY IMPORTED_LOCATION $ENV{STAGING_DIR}/usr/lib/dcv_sub.a)
         list(APPEND LIBRARIES ddplus)
         add_definitions("-DHAVE_DDPLUS=1")
     endif()

