diff --git a/netflix/src/platform/gibbon/netflix.supp b/netflix/src/platform/gibbon/netflix.supp
index b48ef77..2e75f94 100644
--- a/netflix/src/platform/gibbon/netflix.supp
+++ b/netflix/src/platform/gibbon/netflix.supp
@@ -49,13 +49,13 @@
 {
    <insert_a_suppression_name_here>
    Memcheck:Value4
-   obj:/usr/lib/i386-linux-gnu/libstdc++.so.6.0.16
+   obj:libstdc++.so.6.0.16
    obj:*
 }
 {
    <insert_a_suppression_name_here>
    Memcheck:Cond
-   obj:/usr/lib/i386-linux-gnu/libstdc++.so.6.0.16
+   obj:libstdc++.so.6.0.16
    obj:*
 }
 {
@@ -157,7 +157,7 @@
    Memcheck:Param
    writev(vector[...])
    fun:writev
-   obj:/usr/lib/i386-linux-gnu/libX11.so.6.3.0
+   obj:libX11.so.6.3.0
    fun:_XSend
    obj:*
 }
@@ -166,7 +166,7 @@
    Memcheck:Param
    writev(vector[...])
    fun:writev
-   obj:/usr/lib/i386-linux-gnu/libX11.so.6.3.0
+   obj:libX11.so.6.3.0
 }
 {
    <insert_a_suppression_name_here>
diff --git a/partner/graphics/gles2/gibbon_graphics.cmake b/partner/graphics/gles2/gibbon_graphics.cmake
index e23cefc..675c011 100644
--- a/partner/graphics/gles2/gibbon_graphics.cmake
+++ b/partner/graphics/gles2/gibbon_graphics.cmake
@@ -70,8 +70,8 @@ if(GIBBON_GRAPHICS_GL_WSYS STREQUAL "glx")
         endforeach()
 
         # Link against the system's libc, as X11 or OpenGL might require it
-        if(EXISTS /lib/i386-linux-gnu/libc.so.6)
-            gibbon_add_libraries(/lib/i386-linux-gnu/libc.so.6)
+        if(EXISTS libc.so.6)
+            gibbon_add_libraries(libc.so.6)
         endif()
 
     endif()
diff --git a/partner/input/x11/gibbon_input.cmake b/partner/input/x11/gibbon_input.cmake
index db4b2e7..f9f42d7 100644
--- a/partner/input/x11/gibbon_input.cmake
+++ b/partner/input/x11/gibbon_input.cmake
@@ -16,8 +16,8 @@ add_custom_command(OUTPUT ${CMAKE_BINARY_DIR}/include/X11
 add_custom_target(gibbon_x11_links DEPENDS "${CMAKE_BINARY_DIR}/include/X11")
 gibbon_add_dependencies(gibbon_x11_links)
 
-if(EXISTS /lib/i386-linux-gnu/libc.so.6)
-    gibbon_add_libraries(/lib/i386-linux-gnu/libc.so.6)
+if(EXISTS libc.so.6)
+    gibbon_add_libraries(libc.so.6)
 endif()
 
-gibbon_add_libraries(${X11_LIBRARIES})
\ No newline at end of file
+gibbon_add_libraries(${X11_LIBRARIES})
