--- nrd-master/netflix/dpidir/reference/CMakeLists.org	2015-01-31 13:30:52.078495607 +0100
+++ nrd-master/netflix/dpidir/reference/CMakeLists.txt	2015-01-31 13:35:10.246498095 +0100
@@ -14,6 +14,7 @@
 include_directories(${DPI_COMMON}/AudioMixerSoftware)
 include_directories(${DPI_COMMON}/ies)
 include_directories(${DPI_COMMON}/NTBA)
+include_directories($ENV{STAGING_DIR}/usr/include/directfb)
 
 add_definitions(-D__STDC_CONSTANT_MACROS)
 
