--- nrd/netflix/gibbondir/platform/application-manager/ApplicationManagerClient.org	2015-03-02 18:00:02.166559953 +0100
+++ nrd/netflix/gibbondir/platform/application-manager/ApplicationManagerClient.cpp	2015-03-02 18:10:09.590565805 +0100
@@ -7,6 +7,7 @@
 #include <unistd.h>
 #include <pthread.h>
 #include <dlfcn.h>
+#include <bcm_host.h>
 #include "ApplicationManager.h"
 #include "ApplicationManagerClient.h"
 
@@ -61,6 +62,9 @@
 
 ApplicationManagerClient::ApplicationManagerClient(ApplicationManagerCallbacks & cb) : mThread(0), mRunning(true), mCb(cb)
 {
+    // If we start this, we startthe netflix app, this is a singleton!!
+    bcm_host_init();
+
     pthread_create(&mThread, NULL, &ApplicationManagerClient::Task, this);
 }
 
--- a/netflix/src/platform/gibbon/main.cpp	2015-01-26 22:34:21.000000000 +0100
+++ b/netflix/src/platform/gibbon/main.cpp	2015-03-02 19:30:23.338612182 +0100
@@ -7,6 +7,7 @@
  * prohibited.
  */
 
+#include <bcm_host.h>
 #include <nrdbase/config.h>
 #include <gibbon/config.h>
 #include <nrdbase/Configuration.h>
@@ -514,6 +515,8 @@
 # endif
 
 #endif
+    // If we start this, we startthe netflix app, this is a singleton!!
+    bcm_host_init();
 
     int ret;
     {

