--- a/nexus/build/os/linuxkernel/os_flags.inc
+++ b/nexus/build/os/linuxkernel/os_flags.inc
@@ -113,7 +113,7 @@
 # Seperate MIPS and ARM here
 ifneq ($(filter ${B_REFSW_ARCH}, arm-linux), ${B_REFSW_ARCH})
 NEXUS_CFLAGS += -fomit-frame-pointer -G 0 -mno-abicalls -fno-pic -Wa,--trap -DMODULE -mlong-calls
-NEXUS_CFLAGS += -msoft-float -pipe
+NEXUS_CFLAGS += -mhard-float -pipe
 NEXUS_CFLAGS += -DLINUX -D__KERNEL__
 ifeq ($(LINUX_VER_GE_2_6_31),y)
 LINUX_ASM_INC ?= $(LINUX)/arch/mips/include/asm/mach-brcmstb
--- a/BSEAV/build/refsw_inc.mak
+++ b/BSEAV/build/refsw_inc.mak
@@ -214,7 +214,6 @@
 endif
 # -O2 is required for kernel inline functions
 B_REFSW_CFLAGS += -O2
-B_REFSW_CFLAGS += -msoft-float
 B_REFSW_CFLAGS += -pipe
 
 ifeq ($(LINUX_VER_GE_2_6_31),y)
