--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -659,8 +659,8 @@
       }
 
       buffers = g_list_append (buffers, buffer);
-      gst_egl_image_memory_set_orientation (mem,
-          GST_VIDEO_GL_TEXTURE_ORIENTATION_X_NORMAL_Y_FLIP);
+      /*gst_egl_image_memory_set_orientation (mem,
+          GST_VIDEO_GL_TEXTURE_ORIENTATION_X_NORMAL_Y_FLIP);*/
       images = g_list_append (images, gst_egl_image_memory_get_image (mem));
       if (!display)
         display = gst_egl_image_memory_get_display (mem);
