From 9571617bbc5b91337e5a79e9103519e2226b03ca Mon Sep 17 00:00:00 2001
From: Marco Barisione <marco.barisione@collabora.co.uk>
Date: Tue, 28 Jan 2014 15:44:59 +0000
Subject: rpi: change ranks so that omxh264dec can be selected

---
diff --git a/config/rpi/gstomx.conf b/config/rpi/gstomx.conf
index a4b6f26..99fa793 100644
--- a/config/rpi/gstomx.conf
+++ b/config/rpi/gstomx.conf
@@ -2,7 +2,7 @@
 type-name=GstOMXMPEG2VideoDec
 core-name=/opt/vc/lib/libopenmaxil.so
 component-name=OMX.broadcom.video_decode
-rank=256
+rank=257
 in-port-index=130
 out-port-index=131
 hacks=no-component-role
@@ -11,7 +11,7 @@ hacks=no-component-role
 type-name=GstOMXMPEG4VideoDec
 core-name=/opt/vc/lib/libopenmaxil.so
 component-name=OMX.broadcom.video_decode
-rank=256
+rank=257
 in-port-index=130
 out-port-index=131
 hacks=no-component-role
@@ -20,7 +20,7 @@ hacks=no-component-role
 type-name=GstOMXH263Dec
 core-name=/opt/vc/lib/libopenmaxil.so
 component-name=OMX.broadcom.video_decode
-rank=256
+rank=257
 in-port-index=130
 out-port-index=131
 hacks=no-component-role
@@ -29,7 +29,7 @@ hacks=no-component-role
 type-name=GstOMXH264Dec
 core-name=/opt/vc/lib/libopenmaxil.so
 component-name=OMX.broadcom.video_decode
-rank=256
+rank=257
 in-port-index=130
 out-port-index=131
 hacks=no-component-role
@@ -38,7 +38,7 @@ hacks=no-component-role
 type-name=GstOMXTheoraDec
 core-name=/opt/vc/lib/libopenmaxil.so
 component-name=OMX.broadcom.video_decode
-rank=256
+rank=257
 in-port-index=130
 out-port-index=131
 hacks=no-component-role
@@ -47,7 +47,7 @@ hacks=no-component-role
 type-name=GstOMXVP8Dec
 core-name=/opt/vc/lib/libopenmaxil.so
 component-name=OMX.broadcom.video_decode
-rank=256
+rank=257
 in-port-index=130
 out-port-index=131
 hacks=no-component-role
@@ -56,7 +56,7 @@ hacks=no-component-role
 type-name=GstOMXMJPEGDec
 core-name=/opt/vc/lib/libopenmaxil.so
 component-name=OMX.broadcom.video_decode
-rank=256
+rank=257
 in-port-index=130
 out-port-index=131
 hacks=no-component-role
@@ -75,7 +75,7 @@ sink-template-caps=video/x-wmv,wmvversion=(int)3,format=(string){WMV3,WVC1},widt
 type-name=GstOMXH264Enc
 core-name=/opt/vc/lib/libopenmaxil.so
 component-name=OMX.broadcom.video_encode
-rank=256
+rank=257
 in-port-index=200
 out-port-index=201
 hacks=no-component-role
