From 0e0d15b5708c1fe8192af5854b006145246c0fb0 Mon Sep 17 00:00:00 2001
From: Julien Isorce <julien.isorce@collabora.co.uk>
Date: Mon, 27 Jan 2014 17:03:50 +0000
Subject: omxvideodec: populate the most downstream output port on reset

Make seeking work when using egl_render component
---
diff --git a/omx/gstomxvideodec.c b/omx/gstomxvideodec.c
index a24282a..af1e690 100644
--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -1944,6 +1944,7 @@ static gboolean
 gst_omx_video_dec_reset (GstVideoDecoder * decoder, gboolean hard)
 {
   GstOMXVideoDec *self;
+  OMX_ERRORTYPE err = OMX_ErrorNone;
 
   self = GST_OMX_VIDEO_DEC (decoder);
 
@@ -1958,8 +1959,10 @@ gst_omx_video_dec_reset (GstVideoDecoder * decoder, gboolean hard)
   gst_omx_port_set_flushing (self->dec_out_port, 5 * GST_SECOND, TRUE);
 
 #if defined (USE_OMX_TARGET_RPI) && defined (HAVE_GST_EGL)
-  gst_omx_port_set_flushing (self->egl_in_port, 5 * GST_SECOND, TRUE);
-  gst_omx_port_set_flushing (self->egl_out_port, 5 * GST_SECOND, TRUE);
+  if (self->eglimage) {
+    gst_omx_port_set_flushing (self->egl_in_port, 5 * GST_SECOND, TRUE);
+    gst_omx_port_set_flushing (self->egl_out_port, 5 * GST_SECOND, TRUE);
+  }
 #endif
 
   /* Wait until the srcpad loop is finished,
@@ -1972,13 +1975,24 @@ gst_omx_video_dec_reset (GstVideoDecoder * decoder, gboolean hard)
 
   gst_omx_port_set_flushing (self->dec_in_port, 5 * GST_SECOND, FALSE);
   gst_omx_port_set_flushing (self->dec_out_port, 5 * GST_SECOND, FALSE);
-  gst_omx_port_populate (self->dec_out_port);
 
 #if defined (USE_OMX_TARGET_RPI) && defined (HAVE_GST_EGL)
-  gst_omx_port_set_flushing (self->egl_in_port, 5 * GST_SECOND, FALSE);
-  gst_omx_port_set_flushing (self->egl_out_port, 5 * GST_SECOND, FALSE);
+  if (self->eglimage) {
+    gst_omx_port_set_flushing (self->egl_in_port, 5 * GST_SECOND, FALSE);
+    gst_omx_port_set_flushing (self->egl_out_port, 5 * GST_SECOND, FALSE);
+    err = gst_omx_port_populate (self->egl_out_port);
+  } else {
+    err = gst_omx_port_populate (self->dec_out_port);
+  }
+#else
+  err = gst_omx_port_populate (self->dec_out_port);
 #endif
 
+  if (err != OMX_ErrorNone) {
+    GST_WARNING_OBJECT (self, "Failed to populate output port: %s (0x%08x)",
+        gst_omx_error_to_string (err), err);
+  }
+
   /* Start the srcpad loop again */
   self->last_upstream_ts = 0;
   self->eos = FALSE;
--
cgit v0.9.0.2-2-gbebe
