--- a/Source/WebKit/qt/Api/qwebsettings.cpp
+++ b/Source/WebKit/qt/Api/qwebsettings.cpp
@@ -153,8 +153,15 @@
         settings->setAcceleratedCompositingEnabled(value);
 
         bool showDebugVisuals = qgetenv("WEBKIT_SHOW_COMPOSITING_DEBUG_VISUALS") == "1";
-        settings->setShowDebugBorders(showDebugVisuals);
-        settings->setShowRepaintCounter(showDebugVisuals);
+        value = attributes.value(QWebSettings::DebugBorder,
+                                      global->attributes.value(QWebSettings::DebugBorder));
+
+        settings->setShowDebugBorders(showDebugVisuals || value);
+
+        value = attributes.value(QWebSettings::RepaintCounter,
+                                      global->attributes.value(QWebSettings::RepaintCounter));
+
+        settings->setShowRepaintCounter(showDebugVisuals || value);
 #endif
 #if ENABLE(WEBGL)
         value = attributes.value(QWebSettings::WebGLEnabled,
@@ -288,6 +295,10 @@
         settings->setNeedsSiteSpecificQuirks(value);
 
         settings->setUsesPageCache(WebCore::pageCache()->capacity());
+
+        value = attributes.value(QWebSettings::WebSecurityEnabled,
+                                      global->attributes.value(QWebSettings::WebSecurityEnabled));
+        settings->setWebSecurityEnabled(value);
     } else {
         QList<QWebSettingsPrivate*> settings = *::allSettings();
         for (int i = 0; i < settings.count(); ++i)
@@ -556,6 +567,8 @@
     d->attributes.insert(QWebSettings::ScrollAnimatorEnabled, false);
     d->attributes.insert(QWebSettings::CaretBrowsingEnabled, false);
     d->attributes.insert(QWebSettings::NotificationsEnabled, true);
+    d->attributes.insert(QWebSettings::RepaintCounter, false);
+    d->attributes.insert(QWebSettings::DebugBorder, false);
     d->offlineStorageDefaultQuota = 5 * 1024 * 1024;
     d->defaultTextEncoding = QLatin1String("iso-8859-1");
     d->thirdPartyCookiePolicy = AlwaysAllowThirdPartyCookies;
@@ -857,6 +870,11 @@
     WTF::releaseFastMallocFreeMemory();        
 }
 
+void QWebSettings::garbageCollectNow()
+{
+    WebCore::gcController().garbageCollectNow();
+}
+
 /*!
     Sets the maximum number of pages to hold in the memory page cache to \a pages.
 
--- a/Source/WebKit/qt/Api/qwebsettings.h
+++ b/Source/WebKit/qt/Api/qwebsettings.h
@@ -85,7 +85,10 @@
         ScrollAnimatorEnabled,
         CaretBrowsingEnabled,
         NotificationsEnabled,
-        WebAudioEnabled
+        WebAudioEnabled,
+        RepaintCounter,
+        DebugBorder,
+        WebSecurityEnabled
     };
     enum WebGraphic {
         MissingImageGraphic,
@@ -158,6 +161,8 @@
 
     static void clearMemoryCaches();
 
+    static void garbageCollectNow();
+
     static void enablePersistentStorage(const QString& path = QString());
 
     void setThirdPartyCookiePolicy(ThirdPartyCookiePolicy);
