--- a/Tools/MiniBrowser/qt/MiniBrowserApplication.cpp
+++ b/Tools/MiniBrowser/qt/MiniBrowserApplication.cpp
@@ -347,10 +347,10 @@
             m_windowOptions.setRequestedWindowSize(QSize(list.at(0).toInt(), list.at(1).toInt()));
     }
 
-#if HAVE(QTTESTSUPPORT)
-    if (takeOptionFlag(&args, QStringLiteral("--use-test-fonts")))
-        WebKit::QtTestSupport::initializeTestFonts();
-#endif
+// #if HAVE(QTTESTSUPPORT)
+//     if (takeOptionFlag(&args, QStringLiteral("--use-test-fonts")))
+//         WebKit::QtTestSupport::initializeTestFonts();
+// #endif
 
     if (args.contains("-r")) {
         QString listFile = takeOptionValue(&args, "-r");
--- a/Tools/QtTestBrowser/qttestbrowser.cpp
+++ b/Tools/QtTestBrowser/qttestbrowser.cpp
@@ -241,8 +241,8 @@
     }
 #endif
 
-    if (args.contains("-use-test-fonts"))
-        WebKit::QtTestSupport::initializeTestFonts();
+    // if (args.contains("-use-test-fonts"))
+    //     WebKit::QtTestSupport::initializeTestFonts();
 
     if (args.contains("-print-loaded-urls"))
         windowOptions.printLoadedUrls = true;
