--- a/Source/WebKit/qt/WebCoreSupport/ChromeClientQt.cpp	2013-09-05 15:25:27.199783302 +0300
+++ b/Source/WebKit/qt/WebCoreSupport/ChromeClientQt.cpp	2013-09-05 15:34:18.281776981 +0300
@@ -656,7 +656,7 @@ void ChromeClientQt::scheduleCompositing
 ChromeClient::CompositingTriggerFlags ChromeClientQt::allowedCompositingTriggers() const
 {
     if (allowsAcceleratedCompositing())
-        return ThreeDTransformTrigger | CanvasTrigger | AnimationTrigger | AnimatedOpacityTrigger;
+        return ThreeDTransformTrigger | VideoTrigger | CanvasTrigger | AnimationTrigger | AnimatedOpacityTrigger;
 
     return 0;
 }
