--- a/Source/WebCore/loader/FrameLoader.cpp
+++ b/Source/WebCore/loader/FrameLoader.cpp
@@ -2985,6 +2985,11 @@
 
 bool FrameLoader::shouldInterruptLoadForXFrameOptions(const String& content, const KURL& url, unsigned long requestIdentifier)
 {
+    Settings* settings = m_frame->document()->settings();
+
+    if (settings == NULL || settings->webSecurityEnabled() == false)
+	return false;
+ 
     FeatureObserver::observe(m_frame->document(), FeatureObserver::XFrameOptions);
 
     Frame* topFrame = m_frame->tree()->top();
