diff -Naur qt5webkit-5.2.1-orig/Source/WebKit/qt/WebCoreSupport/InitWebCoreQt.cpp qt5webkit-5.2.1/Source/WebKit/qt/WebCoreSupport/InitWebCoreQt.cpp
--- qt5webkit-5.2.1-orig/Source/WebKit/qt/WebCoreSupport/InitWebCoreQt.cpp	2014-03-06 13:29:21.974186444 +0100
+++ qt5webkit-5.2.1/Source/WebKit/qt/WebCoreSupport/InitWebCoreQt.cpp	2014-03-06 13:29:34.530219939 +0100
@@ -49,6 +49,14 @@
 #include <JavaScriptCore/runtime/InitializeThreading.h>
 #include <wtf/MainThread.h>
 
+#include <QDir>
+#include "CookieJarQt.h"
+#if QT_VERSION >= 0x050000
+#include <QStandardPaths>
+#else
+#include <QDesktopServices>
+#endif
+
 namespace WebKit {
 
 static QtStyleFacadeFactoryFunction initCallback = 0;
@@ -109,6 +117,15 @@
         WebCore::memoryCache()->setDeadDecodedDataDeletionInterval(60);
     WebCore::RuntimeEnabledFeatures::setCSSCompositingEnabled(true);
 
+    // initialize CookieJar
+#if QT_VERSION >= 0x050000
+    QString path = QStandardPaths::writableLocation(QStandardPaths::DataLocation);
+#else
+    QString path = QDesktopServices::storageLocation(QDesktopServices::DataLocation);
+#endif
+    QDir().mkpath(path);
+    SharedCookieJarQt::create(path);
+
     initialized = true;
 }
 
--- a/Source/WebKit/qt/WebCoreSupport/QWebPageAdapter.cpp
+++ b/Source/WebKit/qt/WebCoreSupport/QWebPageAdapter.cpp
@@ -111,6 +111,7 @@
 #include <QOpenGLContext>
 #include "GLSharedContext.h"
 #endif
+#include "CookieJarQt.h"
 
 // from text/qfont.cpp
 QT_BEGIN_NAMESPACE
@@ -360,8 +361,14 @@
 
 QNetworkAccessManager* QWebPageAdapter::networkAccessManager()
 {
-    if (!networkManager)
+    if (!networkManager) { 
         networkManager = new QNetworkAccessManager(handle());
+
+        // set the shared cookieJar by default to the network manager
+        SharedCookieJarQt *cookieJar = SharedCookieJarQt::shared();
+        networkManager->setCookieJar(cookieJar);
+        cookieJar->setParent(0);
+    }
     return networkManager;
 }
