From 640e3ecfee1298c47af943ada8b32943dac280cb Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tim-Philipp=20M=C3=BCller?= <tim@centricular.com>
Date: Sat, 18 Apr 2015 12:27:46 +0100
Subject: gst.h: include the new gstprotection.h header

https://bugzilla.gnome.org/show_bug.cgi?id=705991

diff --git a/gst/gst.h b/gst/gst.h
index e4e2a1c..8ebd76c 100644
--- a/gst/gst.h
+++ b/gst/gst.h
@@ -64,6 +64,7 @@
 #include <gst/gstplugin.h>
 #include <gst/gstpoll.h>
 #include <gst/gstpreset.h>
+#include <gst/gstprotection.h>
 #include <gst/gstquery.h>
 #include <gst/gstregistry.h>
 #include <gst/gstsample.h>
-- 
cgit v0.10.2
