--- /dev/null
+++ b/mkspecs/devices/linux-buildroot-g++/qmake.conf
@@ -7,5 +7,11 @@
 
 QMAKE_LIBS             += -lrt -lpthread -ldl
 
+ifeq ($(BR2_BCM_REFSW_VERSION_V15),y)
+QMAKE_INCDIR_OPENGL_ES2 = $(STAGING_DIR)/usr/include/GLES2
+QMAKE_LIBDIR_OPENGL_ES2 = $(STAGING_DIR)/usr/lib
+QMAKE_LIBS_OPENGL_ES2   += -lnexus -lnxpl -lv3ddriver
+endif
+
 include(../common/linux_device_post.conf)
 load(qt_config)
