--- /src/plugins/platforms/eglfs/eglfs.pri
+++ /src/plugins/platforms/eglfs/eglfs.pri
@@ -5,8 +5,8 @@
 
 # Uncomment these to enable the X hooks, allowing to test the platform
 # plugin in a regular X11 environment (as long as EGL is available).
-# EGLFS_PLATFORM_HOOKS_SOURCES += qeglfshooks_x11.cpp
-# LIBS += -lX11 -lX11-xcb -lxcb
+EGLFS_PLATFORM_HOOKS_SOURCES += qeglfshooks_x11.cpp
+LIBS += -lX11 -lX11-xcb -lxcb
 
 # Uncomment these to enable the KMS hooks.
 # EGLFS_PLATFORM_HOOKS_SOURCES += qeglfshooks_kms.cpp
