diff -Naur qt5base-5.2.1-orig/src/gui/painting/qpaintengineex.cpp qt5base-5.2.1/src/gui/painting/qpaintengineex.cpp
--- qt5base-5.2.1-orig/src/gui/painting/qpaintengineex.cpp	2014-03-19 16:25:54.878120695 +0100
+++ qt5base-5.2.1/src/gui/painting/qpaintengineex.cpp	2014-03-19 16:38:55.050576140 +0100
@@ -1088,12 +1088,15 @@
 
 bool QPaintEngineEx::shouldDrawCachedGlyphs(QFontEngine *fontEngine, const QTransform &m) const
 {
-    if (fontEngine->glyphFormat == QFontEngineGlyphCache::Raster_ARGB)
-        return true;
+    // if (fontEngine->glyphFormat == QFontEngineGlyphCache::Raster_ARGB)
+    //     return true;
 
-    qreal pixelSize = fontEngine->fontDef.pixelSize;
-    return (pixelSize * pixelSize * qAbs(m.determinant())) <
-            QT_MAX_CACHED_GLYPH_SIZE * QT_MAX_CACHED_GLYPH_SIZE;
+    // qreal pixelSize = fontEngine->fontDef.pixelSize;
+    // return (pixelSize * pixelSize * qAbs(m.determinant())) <
+    //         QT_MAX_CACHED_GLYPH_SIZE * QT_MAX_CACHED_GLYPH_SIZE;
+
+    // we want to use always cached fonts
+    return true;
 }
 
 QT_END_NAMESPACE
