diff --git a/Source/WebCore/platform/gtk/WaylandCompositorDispmanX.cpp b/Source/WebCore/platform/gtk/WaylandCompositorDispmanX.cpp
index 5b5e969..4a47ced 100644
--- a/Source/WebCore/platform/gtk/WaylandCompositorDispmanX.cpp
+++ b/Source/WebCore/platform/gtk/WaylandCompositorDispmanX.cpp
@@ -33,6 +33,7 @@ struct NestedSurfaceDispmanX : NestedSurface {
         : NestedSurface(compositor)
         , subsurfaceSurface(nullptr)
         , subsurface(nullptr)
+        , queueRedraw(false)
     { }
     // FIXME: Destroy subsurface.
     virtual ~NestedSurfaceDispmanX() { }
@@ -42,6 +43,7 @@ struct NestedSurfaceDispmanX : NestedSurface {
 
     struct wl_surface* subsurfaceSurface;
     struct wl_subsurface* subsurface;
+    bool queueRedraw;
 };
 
 struct NestedBufferDispmanX : NestedBuffer {
@@ -65,8 +67,6 @@ void NestedSurfaceDispmanX::setWidget(GtkWidget* widget)
     fprintf(stderr, "NestedSurfaceDispmanX::setWidget(): acquired subsurface %p, its surface %p\n",
         subsurface, subsurfaceSurface);
     wl_subsurface_set_desync(subsurface);
-
-    gtk_widget_queue_draw(widget);
 }
 
 NestedBuffer* NestedSurfaceDispmanX::createBuffer(struct wl_resource* resource)
@@ -83,6 +83,8 @@ NestedBuffer* NestedSurfaceDispmanX::createBuffer(struct wl_resource* resource)
     buffer->clientBuffer = wl_dispmanx_create_proxy_buffer(static_cast<WaylandCompositorDispmanX*>(compositor)->wl_dispmanx,
         vc_dispmanx_get_handle_from_wl_buffer(resource), width, height,
         vc_dispmanx_get_format_from_wl_buffer(resource));
+
+    queueRedraw = true;
     return buffer;
 }
 
@@ -246,6 +248,12 @@ void WaylandCompositorDispmanX::commitSurface(NestedSurface* surfaceBase, struct
     wl_display_flush_clients(m_display.childDisplay);
     //fprintf(stderr, "\tsurface committed\n");
 
+    if (surface->queueRedraw) {
+        fprintf(stderr, "\tnew buffer -- queueing redraw\n");
+        gtk_widget_queue_draw(surface->widget);
+        surface->queueRedraw = false;
+    }
+
     commitCount++;
     uint64_t commitTime = getCurrentTime();
     if (commitTime - lastCommitPrint >= 5000) {
