diff --git a/Source/WebKit2/GNUmakefile.am b/Source/WebKit2/GNUmakefile.am
index c299b1f..a4bed28 100644
--- a/Source/WebKit2/GNUmakefile.am
+++ b/Source/WebKit2/GNUmakefile.am
@@ -9,12 +9,6 @@ lib_LTLIBRARIES += \
 noinst_LTLIBRARIES += \
 	libWebKit2Platform.la
 
-# We need to make sure plugin process is always built with gtk2, because plugins like flash
-# still use gtk2. Since webcore was built with gtk3, we need to build webcoregtk sources again with gtk2
-noinst_LTLIBRARIES += \
-	libPlatformGtk2.la \
-	libWebCoreGtk2.la
-
 libexec_PROGRAMS += \
 	Programs/WebKitPluginProcess
 endif
@@ -583,68 +577,6 @@ Programs_WebKitWebProcess_LDADD = \
 	$(GTK_UNIX_PRINTING_LIBS) \
 	$(LIBSOUP_LIBS)
 
-libWebCoreGtk2_la_SOURCES = \
-	$(webcoregtk_sources)
-
-libWebCoreGtk2_la_CXXFLAGS = \
-	-fvisibility-inlines-hidden \
-	$(global_cxxflags)
-
-libWebCoreGtk2_la_CFLAGS = \
-	-fvisibility=hidden \
-	$(global_cflags)
-
-libWebCoreGtk2_la_CPPFLAGS = \
-	-DBUILDING_WEBKIT \
-	-DGTK_API_VERSION_2=1 \
-	$(global_cppflags) \
-	$(platform_cppflags) \
-	$(platformgtk_cppflags) \
-	$(webcore_cppflags) \
-	$(webcoregtk_cppflags) \
-	$(javascriptcore_cppflags) \
-	-fno-strict-aliasing \
-	$(COVERAGE_CFLAGS) \
-	$(ENCHANT_CFLAGS) \
-	$(GEOCLUE_CFLAGS) \
-	$(FREETYPE_CFLAGS) \
-	$(GLIB_CFLAGS) \
-	$(GSTREAMER_CFLAGS) \
-	$(GTK2_CFLAGS) \
-	$(LIBSECRET_CFLAGS) \
-	$(LIBSOUP_CFLAGS) \
-	$(LIBXML_CFLAGS) \
-	$(SQLITE3_CFLAGS) \
-	$(UNICODE_CFLAGS) \
-	$(XRENDER_CFLAGS) \
-	$(XT_CFLAGS)
-
-libPlatformGtk2_la_SOURCES = \
-	$(platformgtk_sources)
-
-libPlatformGtk2_la_CXXFLAGS = \
-	-fvisibility-inlines-hidden \
-	$(global_cxxflags)
-
-libPlatformGtk2_la_CFLAGS = \
-	-fvisibility=hidden \
-	$(global_cflags)
-
-libPlatformGtk2_la_CPPFLAGS = \
-	-DBUILDING_WEBKIT \
-	-DGTK_API_VERSION_2=1 \
-	$(global_cppflags) \
-	$(platform_cppflags) \
-	$(platformgtk_cppflags) \
-	$(platform_webcore_cppflags) \
-	$(javascriptcore_cppflags) \
-	-fno-strict-aliasing \
-	$(FREETYPE_CFLAGS) \
-	$(GSTREAMER_CFLAGS) \
-	$(GTK2_CFLAGS) \
-	$(LIBSECRET_CFLAGS) \
-	$(LIBSOUP_CFLAGS)
-
 # Injected Bundle
 injectedbundledir = $(libdir)/webkit2gtk-3.0/injected-bundle
 if ENABLE_WEBKIT2
@@ -682,37 +614,37 @@ plugin_process_for_dep_list += \
 endif
 $(plugin_process_for_dep): $(plugin_process_for_dep_list)
 
-webcore_layer_gtk2_deps = \
+webcore_layer_gtk_deps = \
 	libPlatform.la \
-	libPlatformGtk2.la \
+	libPlatformGtk.la \
 	libWebCorePlatform.la \
-	libWebCoreGtk2.la \
+	libWebCoreGtk.la \
 	libWebCore.la \
 	libWebCoreModules.la
 if ENABLE_SVG
-webcore_layer_gtk2_deps += \
+webcore_layer_gtk_deps += \
 	libWebCoreSVG.la
 endif
 if ENABLE_INDEXED_DATABASE
-webcore_layer_gtk2_deps += \
+webcore_layer_gtk_deps += \
 	libLevelDB.la
 endif
 if USE_OPENGL
-webcore_layer_gtk2_deps += \
+webcore_layer_gtk_deps += \
 	libANGLE.la
 endif
 
-webcore_layer_gtk2_archives = $(foreach lib, $(webcore_layer_gtk2_deps), $(shell echo $(lib) | sed "s/\(.*\)\.la/.libs\/\1.a/"))
+webcore_layer_gtk_archives = $(foreach lib, $(webcore_layer_gtk_deps), $(shell echo $(lib) | sed "s/\(.*\)\.la/.libs\/\1.a/"))
 
-libWebCoreLayerGtk2.a: $(webcore_layer_gtk2_deps)
+libWebCoreLayerGtk.a: $(webcore_layer_gtk_deps)
 	$(AM_V_GEN)
 	$(AM_V_at)$(shell rm -f $@)
 	$(AM_V_at)$(shell find . -name "*.o" > objects_list)
-	$(AM_V_at)$(foreach archive, $(webcore_layer_gtk2_archives), $(shell $(AR) t $(archive) | xargs -n1 basename | xargs -I obj_file grep -F obj_file objects_list | xargs -n50 $(AR) $(AR_FLAGS) $@))
+	$(AM_V_at)$(foreach archive, $(webcore_layer_gtk_archives), $(shell $(AR) t $(archive) | xargs -n1 basename | xargs -I obj_file grep -F obj_file objects_list | xargs -n50 $(AR) $(AR_FLAGS) $@))
 	$(AM_V_at)$(shell rm -f objects_list)
 
 DISTCLEANFILES += \
-	$(top_builddir)/libWebCoreLayerGtk2.a
+	$(top_builddir)/libWebCoreLayerGtk.a
 
 Programs_WebKitPluginProcess_CPPFLAGS = \
 	-include $(srcdir)/Source/WebKit2/WebKit2Prefix.h \
@@ -735,7 +667,6 @@ Programs_WebKitPluginProcess_CPPFLAGS = \
 	-I$(top_builddir)/DerivedSources/WebKit2/include/WebCore \
 	-I$(top_builddir)/DerivedSources/WebKit2/include/WebKit2 \
 	-DBUILDING_WEBKIT \
-	-DGTK_API_VERSION_2=1 \
 	-DENABLE_PLUGIN_PROCESS=1 \
 	-DLIBEXECDIR=\""$(libexecdir)"\" \
 	$(global_cppflags) \
@@ -749,7 +680,7 @@ Programs_WebKitPluginProcess_CPPFLAGS = \
 	$(FREETYPE_CFLAGS) \
 	$(GLIB_CFLAGS) \
 	$(GSTREAMER_CFLAGS) \
-	$(GTK2_CFLAGS) \
+	$(GTK_CFLAGS) \
 	$(LIBSOUP_CFLAGS) \
 	$(UNICODE_CFLAGS)
 
@@ -759,7 +690,7 @@ Programs_WebKitPluginProcess_CFLAGS = \
 	$(CAIRO_CFLAGS) \
 	$(GLIB_CFLAGS) \
 	$(GSTREAMER_CFLAGS) \
-	$(GTK2_CFLAGS) \
+	$(GTK_CFLAGS) \
 	$(LIBSOUP_CFLAGS) \
 	$(UNICODE_CFLAGS)
 
@@ -772,7 +703,7 @@ Programs_WebKitPluginProcess_SOURCES = \
 Programs_WebKitPluginProcess_LDADD = \
 	-lpthread \
 	libjavascriptcoregtk-@WEBKITGTK_API_MAJOR_VERSION@.@WEBKITGTK_API_MINOR_VERSION@.la \
-	libWebCoreLayerGtk2.a \
+	libWebCoreLayerGtk.a \
 	libWebKit2Platform.la \
 	$(CAIRO_LIBS) \
 	$(COVERAGE_LDFLAGS) \
@@ -781,7 +712,7 @@ Programs_WebKitPluginProcess_LDADD = \
 	$(GEOCLUE_LIBS) \
 	$(GLIB_LIBS) \
 	$(GSTREAMER_LIBS) \
-	$(GTK2_LIBS) \
+	$(GTK_LIBS) \
 	$(JPEG_LIBS) \
 	$(LIBSECRET_LIBS) \
 	$(LIBSOUP_LIBS) \
diff --git a/Source/autotools/FindDependencies.m4 b/Source/autotools/FindDependencies.m4
index 4be8a8d..9c8c04f 100644
--- a/Source/autotools/FindDependencies.m4
+++ b/Source/autotools/FindDependencies.m4
@@ -501,11 +501,6 @@ if test "$enable_webkit2" = "yes"; then
         AC_MSG_ERROR([WebKit2 requires GTK+ 3.x, use --with-gtk=3.0])
     fi
 
-    # Make sure we have GTK+ 2.x to build the plugin process.
-    PKG_CHECK_MODULES(GTK2, gtk+-2.0 >= gtk2_required_version)
-    AC_SUBST(GTK2_CFLAGS)
-    AC_SUBST(GTK2_LIBS)
-
     # Check Unix printing
     PKG_CHECK_MODULES(GTK_UNIX_PRINTING, gtk+-unix-print-3.0, [have_gtk_unix_printing=yes], [have_gtk_unix_printing=no])
     AC_SUBST(GTK_UNIX_PRINTING_CFLAGS)
