--- a/thirdparty/khronos/EGL/eglplatform.h
+++ b/thirdparty/khronos/EGL/eglplatform.h
@@ -111,6 +111,8 @@
 
 #elif defined(__unix__)
 
+#if defined(HAVE_X11)
+
 /* X11 (tentative)  */
 #include <X11/Xlib.h>
 #include <X11/Xutil.h>
@@ -120,6 +122,14 @@
 typedef Window   EGLNativeWindowType;
 
 #else
+
+typedef void *EGLNativeDisplayType;
+typedef void *EGLNativePixmapType;
+typedef void *EGLNativeWindowType;
+
+#endif
+
+#else
 #error "Platform not recognized"
 #endif
 
