--- a/libgupnp/gupnp-device-info.c
+++ b/libgupnp/gupnp-device-info.c
@@ -149,6 +149,9 @@
         case PROP_URL_BASE:
                 g_value_set_boxed (value, info->priv->url_base);
                 break;
+        case PROP_DOCUMENT:
+                g_value_set_object (value, info->priv->doc);
+                break;
         default:
                 G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
                 break;
@@ -334,7 +337,7 @@
                                       "The XML document related to this "
                                       "device",
                                       GUPNP_TYPE_XML_DOC,
-                                      G_PARAM_WRITABLE |
+                                      G_PARAM_READWRITE |
                                       G_PARAM_CONSTRUCT_ONLY |
                                       G_PARAM_STATIC_NAME |
                                       G_PARAM_STATIC_NICK |
