diff -Naur nexus-14.2.1-old/nexus/lib/ipc/build/Makefile nexus-14.2.1-new/nexus/lib/ipc/build/Makefile
--- nexus-14.2.1-old/nexus/lib/ipc/build/Makefile	2014-08-07 16:07:36.545697551 +0530
+++ nexus-14.2.1-new/nexus/lib/ipc/build/Makefile	2014-08-07 16:05:03.145698682 +0530
@@ -118,7 +118,7 @@
 LDFLAGS += -lpthread
 endif
 
-all: ${ODIR}/flag apps
+all: ${ODIR}/flag apps ipclib
 
 .NOTPARALLEL: 
 
@@ -139,6 +139,19 @@
 	@echo $(CC) -o $@
 	$(Q_)$(CC) -o $@ $(filter-out %.h, $^) $(CFLAGS) $(LDFLAGS)
 
+ipclib:
+	$(Q_)$(CC) -fPIC -g -c -Wall $(NEXUS_TOP)/lib/ipc/bipc_client.c $(CFLAGS) $(LDFLAGS)
+	$(Q_)$(CC) -fPIC -g -c -Wall $(NEXUS_TOP)/lib/ipc/bipc_server.c $(CFLAGS) $(LDFLAGS)
+	$(Q_)$(CC) -fPIC -g -c -Wall $(NEXUS_TOP)/lib/ipc/bipc_util.c $(CFLAGS) $(LDFLAGS)
+	$(CC) -shared -Wl,-soname,libb_ipc.so -o libb_ipc.so bipc_client.o bipc_server.o bipc_util.o -lc
+	$(CP) libb_ipc.so $(NEXUS_BIN_DIR)/
+
+ipclib-clean:
+	$(RM) $(NEXUS_TOP)/lib/ipc/build/bipc_client.o
+	$(RM) $(NEXUS_TOP)/lib/ipc/build/bipc_server.o
+	$(RM) $(NEXUS_TOP)/lib/ipc/build/bipc_util.o
+	$(RM) $(NEXUS_BIN_DIR)/libb_ipc.so
+
 debug: apps
 	echo "set args ${CLIENT}" >.gdbinit
 	echo "r" >>.gdbinit
