diff --git a/audfilter/src/gst_brcm_aud_filter.c b/audfilter/src/gst_brcm_aud_filter.c
index 556cbd6..6b856a0 100755
--- a/audfilter/src/gst_brcm_aud_filter.c
+++ b/audfilter/src/gst_brcm_aud_filter.c
@@ -47,6 +47,9 @@
 #include <stdio.h>
 #include <stdlib.h>
 
+#include <strings.h>
+#include <nexus_config.h>
+
 #include "nexus_platform.h"
 #include "nexus_core_utils.h"
 
@@ -590,7 +593,7 @@ static void gst_brcm_aud_filter_get_property( GObject *object, guint prop_id, GV
         break;       
     }
      case PROP_FIFO_SIZE:
-        g_value_set_uint(value, (uint) filter->playpump_fifoSize);
+        g_value_set_uint(value, (unsigned int) filter->playpump_fifoSize);
         break;
     default:
         G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
diff --git a/audfilter/src/gst_brcm_p_aac_repack.c b/audfilter/src/gst_brcm_p_aac_repack.c
index 6b473ad..80b7798 100755
--- a/audfilter/src/gst_brcm_p_aac_repack.c
+++ b/audfilter/src/gst_brcm_p_aac_repack.c
@@ -46,6 +46,9 @@
 #include <gst/gst.h>
 #include <stdio.h>
 
+#include <strings.h>
+#include <nexus_config.h>
+
 #include "nexus_platform.h"
 #include "nexus_core_utils.h"
 
diff --git a/audiodecode/src/gst_brcm_audio_decoder.c b/audiodecode/src/gst_brcm_audio_decoder.c
index 16f1d12..eff560e 100755
--- a/audiodecode/src/gst_brcm_audio_decoder.c
+++ b/audiodecode/src/gst_brcm_audio_decoder.c
@@ -48,6 +48,9 @@
 #include <gst/gst.h>
 #include <stdio.h>
 
+#include <strings.h>
+#include <nexus_config.h>
+
 #include "bmedia_types.h"
 #include "gst_brcm_audio_decoder.h"
 
diff --git a/audiosink/src/gst_brcm_audio_sink.c b/audiosink/src/gst_brcm_audio_sink.c
index 2d73326..5212109 100755
--- a/audiosink/src/gst_brcm_audio_sink.c
+++ b/audiosink/src/gst_brcm_audio_sink.c
@@ -46,6 +46,9 @@
 #include <gst/gst.h>
 #include <stdio.h>
 
+#include <strings.h>
+#include <nexus_config.h>
+
 #include "gst_brcm_audio_sink.h"
 
 #include "bstd.h"
diff --git a/gfxsink/src/gst_brcm_gfx_sink.c b/gfxsink/src/gst_brcm_gfx_sink.c
index 6a6add5..87d5f4c 100755
--- a/gfxsink/src/gst_brcm_gfx_sink.c
+++ b/gfxsink/src/gst_brcm_gfx_sink.c
@@ -48,6 +48,9 @@
 #include <stdio.h>
 #include <gst/video/video.h>
 
+#include <strings.h>
+#include <nexus_config.h>
+
 #include "gst_brcm_gfx_sink.h"
 
 #include "bstd.h"
diff --git a/mp3swdecode/src/gst_mp3_decoder.c b/mp3swdecode/src/gst_mp3_decoder.c
index 023e26c..49db2b7 100755
--- a/mp3swdecode/src/gst_mp3_decoder.c
+++ b/mp3swdecode/src/gst_mp3_decoder.c
@@ -50,6 +50,9 @@
 #include <gst/audio/audio.h>
 #endif
 
+#include <strings.h>
+#include <nexus_config.h>
+
 #include <stdio.h>
 
 #include "bmedia_types.h"
diff --git a/pcmsink/src/gst_pcm_sink.c b/pcmsink/src/gst_pcm_sink.c
index 1d3fe89..167b507 100755
--- a/pcmsink/src/gst_pcm_sink.c
+++ b/pcmsink/src/gst_pcm_sink.c
@@ -49,6 +49,9 @@
 #include <gst/audio/audio.h>
 #endif
 
+#include <strings.h>
+#include <nexus_config.h>
+
 #include <stdio.h>
 
 #include "gst_pcm_sink.h"
diff --git a/playersinkbin/src/gstplayersinkbin.c b/playersinkbin/src/gstplayersinkbin.c
index e306230..a0c567d 100755
--- a/playersinkbin/src/gstplayersinkbin.c
+++ b/playersinkbin/src/gstplayersinkbin.c
@@ -794,7 +794,7 @@ void gst_player_sinkbin_video_buffer_underflow_callback (GstElement *pSinkElemen
     PBIN_UNUSED(value2);
 
     GstPlayerSinkBin *playersinkbin = GST_PLAYER_SINKBIN (data);
-    uint fifoDepth, fifoSize;
+    unsigned int fifoDepth, fifoSize;
     gboolean is_music_channel = FALSE; 
 
     g_object_get(playersinkbin->demux,"is-music-channel",&is_music_channel,NULL);
@@ -815,7 +815,7 @@ void gst_player_sinkbin_audio_buffer_underflow_callback (GstElement *pSinkElemen
     PBIN_UNUSED(value2);
 
     GstPlayerSinkBin *playersinkbin = GST_PLAYER_SINKBIN (data);
-    uint fifoDepth, fifoSize;
+    unsigned int fifoDepth, fifoSize;
     if (playersinkbin->is_live) {
         g_object_get(playersinkbin->demux,"fifo-depth",&fifoDepth,NULL);
         g_object_get(playersinkbin->demux,"fifo-size",&fifoSize,NULL);
@@ -1242,21 +1242,21 @@ gst_player_sinkbin_get_property (GObject * object, guint prop_id,
         break;
     case PROP_FIFO_SIZE:
        {
-            uint fifoSize;
+            unsigned int fifoSize;
             g_object_get(playersinkbin->demux,"fifo-size",&fifoSize,NULL);
             g_value_set_uint(value,fifoSize);
             break;
         }
     case PROP_FIFO_DEPTH:
        {
-            uint fifoDepth;
+            unsigned int fifoDepth;
             g_object_get(playersinkbin->demux,"fifo-depth",&fifoDepth,NULL);
             g_value_set_uint(value,fifoDepth);
             break;
         }
     case PROP_FIFO_PRE_BUF_MIN:
        {
-            uint fifo_pre_buf_min;
+            unsigned int fifo_pre_buf_min;
             g_object_get(playersinkbin->demux,"fifo-pre-buf-min",&fifo_pre_buf_min,NULL);
             g_value_set_uint(value,fifo_pre_buf_min);
             break;
@@ -1386,14 +1386,14 @@ gst_player_sinkbin_set_property (GObject * object, guint prop_id,
         }
 		case PROP_FIFO_SIZE:
        {
-            uint fifoSize  = g_value_get_uint(value);
+            unsigned int fifoSize  = g_value_get_uint(value);
             g_object_set(playersinkbin->demux,"fifo-size",fifoSize,NULL);
             break;
         }
 
 		case PROP_FIFO_PRE_BUF_MIN:
        {
-            uint fifo_pre_buf_min  = g_value_get_uint(value);
+            unsigned int fifo_pre_buf_min  = g_value_get_uint(value);
             g_object_set(playersinkbin->demux,"fifo-pre-buf-min",fifo_pre_buf_min,NULL);
             break;
         }
diff --git a/systemclock/src/gst_system_clock.c b/systemclock/src/gst_system_clock.c
index 3867834..d48debe 100755
--- a/systemclock/src/gst_system_clock.c
+++ b/systemclock/src/gst_system_clock.c
@@ -43,6 +43,9 @@
 #include <stdio.h>
 #include <stdint.h>
 
+#include <strings.h>
+#include <nexus_config.h>
+
 #include "nexus_platform.h"
 #include "nexus_core_utils.h"
 
diff --git a/videodecode/src/gst_brcm_video_decoder.c b/videodecode/src/gst_brcm_video_decoder.c
index b36d439..dc753c5 100755
--- a/videodecode/src/gst_brcm_video_decoder.c
+++ b/videodecode/src/gst_brcm_video_decoder.c
@@ -50,6 +50,9 @@
 #include <stdio.h>
 #include <stdlib.h>
 
+#include <strings.h>
+#include <nexus_config.h>
+
 #include "bmedia_types.h"
 #include "gst_brcm_video_decoder.h"
 #include "bkni.h"
diff --git a/videosink/src/gst_brcm_video_sink.c b/videosink/src/gst_brcm_video_sink.c
index ee7bdd0..46f0558 100755
--- a/videosink/src/gst_brcm_video_sink.c
+++ b/videosink/src/gst_brcm_video_sink.c
@@ -46,6 +46,9 @@
 #include <gst/gst.h>
 #include <stdio.h>
 
+#include <strings.h>
+#include <nexus_config.h>
+
 #include "gst_brcm_video_sink.h"
 
 #include "bstd.h"
diff --git a/videosink/src/video_sink_properties.c b/videosink/src/video_sink_properties.c
index 1974c29..7cb0a82 100755
--- a/videosink/src/video_sink_properties.c
+++ b/videosink/src/video_sink_properties.c
@@ -47,6 +47,9 @@
 #include <stdio.h>
 #include <stdlib.h>
 
+#include <strings.h>
+#include <nexus_config.h>
+
 #include "gst_brcm_video_sink.h"
 
 #include "bstd.h"
diff --git a/vidfilter/src/gst_brcm_p_h264_repack.c b/vidfilter/src/gst_brcm_p_h264_repack.c
index 780afbc..efcb9d7 100755
--- a/vidfilter/src/gst_brcm_p_h264_repack.c
+++ b/vidfilter/src/gst_brcm_p_h264_repack.c
@@ -46,6 +46,9 @@
 #include <gst/gst.h>
 #include <stdio.h>
 
+#include <strings.h>
+#include <nexus_config.h>
+
 #include "nexus_platform.h"
 #include "nexus_core_utils.h"
 
diff --git a/vidfilter/src/gst_brcm_vid_filter.c b/vidfilter/src/gst_brcm_vid_filter.c
index 41bc959..4f1f87c 100755
--- a/vidfilter/src/gst_brcm_vid_filter.c
+++ b/vidfilter/src/gst_brcm_vid_filter.c
@@ -46,6 +46,9 @@
 #include <gst/gst.h>
 #include <stdio.h>
 
+#include <strings.h>
+#include <nexus_config.h>
+
 #include "nexus_platform.h"
 #include "nexus_core_utils.h"
 
