--- a/CMakeLists.txt	2015-04-26 12:16:34.329203182 +0200
+++ b/CMakeLists.txt	2015-04-26 14:06:49.217482666 +0200
@@ -2,8 +2,8 @@
 
 project(vmcs_host_apps)
 
-set(BUILD_MMAL TRUE)
-set(BUILD_MMAL_APPS TRUE)
+set(BUILD_MMAL FALSE)
+set(BUILD_MMAL_APPS FALSE)
 set(vmcs_root ${PROJECT_SOURCE_DIR})
 get_filename_component(VIDEOCORE_ROOT . ABSOLUTE)
 
--- a/host_applications/linux/CMakeLists.txt	2015-04-17 22:13:06.000000000 +0200
+++ b/host_applications/linux/CMakeLists.txt	2015-04-26 14:48:00.721587089 +0200
@@ -1,13 +1,13 @@
 # linux apps
 
 add_subdirectory(libs/bcm_host)
-add_subdirectory(apps/gencmd)
-add_subdirectory(apps/tvservice)
-add_subdirectory(apps/vcmailbox)
-add_subdirectory(apps/raspicam)
-add_subdirectory(libs/sm)
-add_subdirectory(apps/smem)
-add_subdirectory(libs/debug_sym)
+#add_subdirectory(apps/gencmd)
+#add_subdirectory(apps/tvservice)
+#add_subdirectory(apps/vcmailbox)
+#add_subdirectory(apps/raspicam)
+#add_subdirectory(libs/sm)
+#add_subdirectory(apps/smem)
+#add_subdirectory(libs/debug_sym)
 
 if(ALL_APPS)
  add_subdirectory(apps/vcdbg)
