diff --git a/interface/vcos/pthreads/CMakeLists.txt b/interface/vcos/pthreads/CMakeLists.txt
index 1d81ca3..d6cd415 100644
--- a/interface/vcos/pthreads/CMakeLists.txt
+++ b/interface/vcos/pthreads/CMakeLists.txt
@@ -33,6 +33,14 @@ set (SOURCES
    ../generic/vcos_generic_blockpool.c
 )
 
+if (BUILD_WAYLAND)
+wayland_add_protocol_server(
+    SOURCES
+    ../../../interface/wayland/dispmanx.xml
+    dispmanx
+)
+endif ()
+
 if (VCOS_PTHREADS_BUILD_SHARED)
    add_library (vcos SHARED ${SOURCES})
    target_link_libraries (vcos pthread dl rt)
diff --git a/interface/vmcs_host/CMakeLists.txt b/interface/vmcs_host/CMakeLists.txt
index 35150ed..22949a8 100755
--- a/interface/vmcs_host/CMakeLists.txt
+++ b/interface/vmcs_host/CMakeLists.txt
@@ -17,14 +17,6 @@ set(VCHOSTIF_SOURCE
 #            ${VMCS_TARGET}/vmcs_main.c
 #  vc_vchi_haud.c
 
-if (BUILD_WAYLAND)
-wayland_add_protocol_server(
-    VCHOSTIF_SOURCE
-    ../../interface/wayland/dispmanx.xml
-    dispmanx
-)
-endif ()
-
 add_library(vchostif ${VCHOSTIF_SOURCE})
 
 #add_library(bufman            vc_vchi_bufman.c            )
diff --git a/interface/vmcs_host/vc_dispmanx.h b/interface/vmcs_host/vc_dispmanx.h
index acbdbb9..e2553c5 100755
--- a/interface/vmcs_host/vc_dispmanx.h
+++ b/interface/vmcs_host/vc_dispmanx.h
@@ -38,7 +38,7 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #ifdef BUILD_WAYLAND
 #include <wayland-server.h>
-#include "interface/vmcs_host/wayland-dispmanx-server-protocol.h"
+#include "interface/vcos/pthreads/wayland-dispmanx-server-protocol.h"
 #endif
 
 #ifdef __cplusplus
diff --git a/interface/vmcs_host/vc_vchi_dispmanx.h b/interface/vmcs_host/vc_vchi_dispmanx.h
index f0bae30..8c44c58 100644
--- a/interface/vmcs_host/vc_vchi_dispmanx.h
+++ b/interface/vmcs_host/vc_vchi_dispmanx.h
@@ -69,7 +69,7 @@ typedef struct {
 #ifdef BUILD_WAYLAND
 /* XXX: This should be in a private header that can be included from EGL and vc_* */
 #include <wayland-server.h>
-#include "interface/vmcs_host/wayland-dispmanx-server-protocol.h"
+#include "interface/vcos/pthreads/wayland-dispmanx-server-protocol.h"
 struct wl_dispmanx_server_buffer {
 	struct wl_resource *resource;
 	struct wl_dispmanx *dispmanx;
