--- a/netflix/src/platform/gibbon/InputManager.h
+++ b/netflix/src/platform/gibbon/InputManager.h
@@ -93,6 +93,19 @@
 
 #elif defined(GIBBON_INPUT_NULL)
     netflix::base::ConditionVariable mCondition;
+#elif defined(GIBBON_INPUT_METROLOGICAL)
+public: 
+    struct QueueInfo
+    {
+      InputEvent::Type type;
+      KeyEvent::Key keyCode;
+      unsigned int keyValue;
+    };
+
+private:
+    std::list<QueueInfo> mInputQueue;
+    netflix::base::ConditionVariable mCondition;
+protected:
 #elif defined(GIBBON_INPUT_X11)
     int mPipe[2];
     shared_ptr<InputManagerX11Private> mPriv;

